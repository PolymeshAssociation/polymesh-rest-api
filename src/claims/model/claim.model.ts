/* istanbul ignore file */

import { ApiProperty } from '@nestjs/swagger';
import { Claim, Identity } from '@polymathnetwork/polymesh-sdk/types';

import { FromEntity } from '~/common/decorators/transformation';

export class ClaimModel<T = Claim> {
  @ApiProperty({
    type: 'string',
    description: 'DID of the target Identity',
    example: '0x0600000000000000000000000000000000000000000000000000000000000000',
  })
  @FromEntity()
  readonly target: Identity;

  @ApiProperty({
    type: 'string',
    description: 'DID of the issuer Identity',
    example: '0x0600000000000000000000000000000000000000000000000000000000000000',
  })
  @FromEntity()
  readonly issuer: Identity;

  @ApiProperty({
    type: 'string',
    description: 'Date when the Claim was issued',
    example: new Date('10/14/1987').toISOString(),
  })
  readonly issuedAt: Date;

  @ApiProperty({
    type: 'string',
    nullable: true,
    description: 'Expiry date of the Claim',
    example: new Date('10/14/1987').toISOString(),
  })
  readonly expiry: Date | null;

  @ApiProperty({
    description: 'Details of the Claim containing type and scope',
    example: {
      type: 'Accredited',
      scope: {
        type: 'Identity',
        value: '0x6'.padEnd(66, '1a'),
      },
    },
  })
  readonly claim: T;

  constructor(model: ClaimModel) {
    Object.assign(this, model);
  }
}
