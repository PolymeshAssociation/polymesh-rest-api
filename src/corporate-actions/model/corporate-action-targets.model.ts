/* istanbul ignore file */

import { ApiProperty } from '@nestjs/swagger';
import { Identity, TargetTreatment } from '@polymathnetwork/polymesh-sdk/types';

import { FromMaybeEntityArray } from '~/common/decorators/transformation';

export class CorporateActionTargetsModel {
  @ApiProperty({
    description: 'Indicates how the `identities` are to be treated',
    type: 'string',
    enum: TargetTreatment,
    example: TargetTreatment.Include,
  })
  readonly treatment: TargetTreatment;

  @ApiProperty({
    description:
      'Determines which Identities will be affected by the Corporate Action. If the value of `treatment` is `Include`, then all Identities in this array will be affected. Otherwise, every Asset holder Identity **EXCEPT** for the ones in this array will be affected',
    type: 'string',
    isArray: true,
    example: [
      '0x0600000000000000000000000000000000000000000000000000000000000000',
      '0x0611111111111111111111111111111111111111111111111111111111111111',
    ],
  })
  @FromMaybeEntityArray()
  readonly identities: Identity[];

  constructor(model: CorporateActionTargetsModel) {
    Object.assign(this, model);
  }
}
