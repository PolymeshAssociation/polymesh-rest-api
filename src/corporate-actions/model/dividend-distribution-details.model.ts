import { ApiProperty } from '@nestjs/swagger';
import { BigNumber } from '@polymathnetwork/polymesh-sdk';
import { DividendDistribution } from '@polymathnetwork/polymesh-sdk/types';

import { FromBigNumber, FromEntity } from '~/common/decorators/transformation';

export class DistributionWithDetailsModel {
  @ApiProperty({
    description: 'Details of the Dividend Distribution',
  })
  @FromEntity()
  distribution: DividendDistribution;

  @ApiProperty({
    description: 'Number of remaining funds',
    type: 'string',
    example: '1000',
  })
  @FromBigNumber()
  remainingFunds: BigNumber;

  @ApiProperty({
    description:
      'Indicates whether the unclaimed funds have been reclaimed by the Corporate Actions Agent',
    type: 'boolean',
    example: false,
  })
  fundsReclaimed: boolean;

  constructor(model: DistributionWithDetailsModel) {
    Object.assign(this, model);
  }
}
