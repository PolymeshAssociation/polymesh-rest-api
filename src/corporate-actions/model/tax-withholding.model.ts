/* istanbul ignore file */

import { ApiProperty } from '@nestjs/swagger';
import { BigNumber } from '@polymathnetwork/polymesh-sdk';
import { Identity } from '@polymathnetwork/polymesh-sdk/internal';

import { FromBigNumber, FromEntity } from '~/common/decorators/transformation';

export class TaxWithholdingModel {
  @ApiProperty({
    description: 'DID for which the tax withholding percentage is overridden',
    type: 'string',
    example: '0x0600000000000000000000000000000000000000000000000000000000000000',
  })
  @FromEntity()
  readonly identity: Identity;

  @ApiProperty({
    description: 'Tax withholding percentage (from 0 to 100)',
    type: 'string',
    example: '67.25',
  })
  @FromBigNumber()
  readonly percentage: BigNumber;

  constructor(model: TaxWithholdingModel) {
    Object.assign(this, model);
  }
}
