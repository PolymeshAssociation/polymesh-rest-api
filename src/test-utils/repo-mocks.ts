/* istanbul ignore file */

import { createMock } from '@golevelup/ts-jest';
import { ValueProvider } from '@nestjs/common';

import { ApiKeyRepo } from '~/auth/repos/api-key.repo';
import { UsersRepo } from '~/users/repo/user.repo';

export const mockApiKeyRepoProvider: ValueProvider<ApiKeyRepo> = {
  provide: ApiKeyRepo,
  useValue: createMock<ApiKeyRepo>(),
};

export const mockUserRepoProvider: ValueProvider<UsersRepo> = {
  provide: UsersRepo,
  useValue: createMock<UsersRepo>(),
};

export class MockQueryBuilder {
  update = jest.fn().mockReturnThis();
  set = jest.fn().mockReturnThis();
  where = jest.fn().mockReturnThis();
  execute = jest.fn().mockResolvedValue({ affected: 1 });
}

/**
 * mocks TypeORM repository
 */
export class MockPostgresRepository {
  findOneBy = jest.fn();
  save = jest.fn();
  delete = jest.fn();
  create = jest.fn();
  update = jest.fn();
  find = jest.fn();
  createQueryBuilder = jest.fn().mockReturnValue(new MockQueryBuilder());
}
