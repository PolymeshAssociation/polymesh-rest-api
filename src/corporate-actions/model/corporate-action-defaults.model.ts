/* istanbul ignore file */

import { ApiProperty } from '@nestjs/swagger';
import { BigNumber } from '@polymathnetwork/polymesh-sdk';
import { Type } from 'class-transformer';

import { FromBigNumber } from '~/common/decorators/transformation';
import { CorporateActionTargetsModel } from '~/corporate-actions/model/corporate-action-targets.model';
import { TaxWithholdingModel } from '~/corporate-actions/model/tax-withholding.model';

export class CorporateActionDefaultsModel {
  @ApiProperty({
    description: 'Identities that will be affected by the Corporate Actions',
    type: CorporateActionTargetsModel,
  })
  @Type(() => CorporateActionTargetsModel)
  readonly targets: CorporateActionTargetsModel;

  @ApiProperty({
    description:
      "Tax withholding percentage(0-100) that applies to Identities that don't have a specific percentage assigned to them",
    type: 'string',
    example: '25',
  })
  @FromBigNumber()
  readonly defaultTaxWithholding: BigNumber;

  @ApiProperty({
    description:
      'List of Identities and the specific tax withholding percentage that should apply to them. This takes precedence over `defaultTaxWithholding`',
    type: TaxWithholdingModel,
    isArray: true,
  })
  @Type(() => TaxWithholdingModel)
  readonly taxWithholdings: TaxWithholdingModel[];

  constructor(model: CorporateActionDefaultsModel) {
    Object.assign(this, model);
  }
}
