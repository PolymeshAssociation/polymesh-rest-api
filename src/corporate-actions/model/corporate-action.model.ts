/* istanbul ignore file */

import { ApiProperty } from '@nestjs/swagger';
import { BigNumber } from '@polymathnetwork/polymesh-sdk';

import { FromBigNumber } from '~/common/decorators/transformation';
import { CorporateActionDefaultsModel } from '~/corporate-actions/model/corporate-action-defaults.model';

export class CorporateActionModel extends CorporateActionDefaultsModel {
  @ApiProperty({
    description: 'ID of the Corporate Action',
    type: 'string',
    example: '1',
  })
  @FromBigNumber()
  readonly id: BigNumber;

  @ApiProperty({
    description: 'Ticker of the Asset',
    type: 'string',
    example: 'TICKER',
  })
  readonly ticker: string;

  @ApiProperty({
    description: 'Date at which the Corporate Action was created',
    type: 'string',
    example: new Date('10/14/1987').toISOString(),
  })
  readonly declarationDate: Date;

  @ApiProperty({
    description: 'Brief description of the Corporate Action',
    type: 'string',
    example: 'Corporate Action description',
  })
  readonly description: string;

  constructor(model: CorporateActionModel) {
    const { targets, defaultTaxWithholding, taxWithholdings, ...rest } = model;
    super({ targets, defaultTaxWithholding, taxWithholdings });

    Object.assign(this, rest);
  }
}
