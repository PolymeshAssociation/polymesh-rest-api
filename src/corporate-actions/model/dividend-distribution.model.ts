/* istanbul ignore file */

import { ApiProperty } from '@nestjs/swagger';
import { BigNumber } from '@polymathnetwork/polymesh-sdk';
import { Type } from 'class-transformer';

import { FromBigNumber } from '~/common/decorators/transformation';
import { CorporateActionModel } from '~/corporate-actions/model/corporate-action.model';
import { PortfolioIdentifierModel } from '~/portfolios/models/portfolio-identifier.model';

export class DividendDistributionModel extends CorporateActionModel {
  @ApiProperty({
    description: 'Portfolio from which the Dividends are distributed',
    type: PortfolioIdentifierModel,
  })
  @Type(() => PortfolioIdentifierModel)
  readonly origin: PortfolioIdentifierModel;

  @ApiProperty({
    description: 'Ticker of the currency in which Dividends are distributed',
    type: 'string',
    example: 'TICKER',
  })
  readonly currency: string;

  @ApiProperty({
    description: "Amount of `currency` to pay for each Asset holders' share",
    type: 'string',
    example: '100',
  })
  @FromBigNumber()
  readonly perShare: BigNumber;

  @ApiProperty({
    description: 'Maximum amount of `currency` to be distributed',
    type: 'string',
    example: '1000',
  })
  @FromBigNumber()
  readonly maxAmount: BigNumber;

  @ApiProperty({
    description:
      'Date after which Dividends can no longer be paid/reclaimed. A null value means the Distribution never expires',
    type: 'string',
    example: new Date('10/14/1987').toISOString(),
    nullable: true,
  })
  readonly expiryDate: null | Date;

  @ApiProperty({
    description: 'Date starting from which dividends can be paid/reclaimed',
    type: 'string',
    example: new Date('10/14/1987').toISOString(),
  })
  readonly paymentDate: Date;

  @ApiProperty({
    description: 'Amount of remaining funds',
    type: 'string',
    example: '1000',
  })
  @FromBigNumber()
  readonly remainingFunds: BigNumber;

  @ApiProperty({
    description:
      'Indicates whether the unclaimed funds have been reclaimed by the Corporate Actions Agent',
    type: 'boolean',
    example: false,
  })
  readonly fundsReclaimed: boolean;

  constructor(model: DividendDistributionModel) {
    const {
      id,
      ticker,
      declarationDate,
      description,
      targets,
      taxWithholdings,
      defaultTaxWithholding,
      ...rest
    } = model;

    super({
      id,
      ticker,
      declarationDate,
      description,
      targets,
      taxWithholdings,
      defaultTaxWithholding,
    });

    Object.assign(this, rest);
  }
}
