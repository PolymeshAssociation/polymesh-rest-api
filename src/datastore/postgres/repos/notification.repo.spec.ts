import { createMock } from '@golevelup/ts-jest';
import { when } from 'jest-when';
import { Repository } from 'typeorm';

import { Notification } from '~/datastore/postgres/entities/notification.entity';
import { PostgresNotificationRepo } from '~/datastore/postgres/repos/notification.repo';
import { NotificationModel } from '~/notifications/model/notification.model';
import { NotificationRepo } from '~/notifications/repo/notifications.repo';
import { MockPostgresRepository } from '~/test-utils/repo-mocks';

describe(`PostgresNotificationRepo ${NotificationRepo.type} test suite`, () => {
  const mockRepository = new MockPostgresRepository();
  const repo = new PostgresNotificationRepo(mockRepository as unknown as Repository<Notification>);

  const mockNotification = createMock<NotificationModel>({
    id: 1,
  });

  mockRepository.save.mockImplementation(notification => {
    notification.id = 1;
    return mockNotification;
  });

  mockRepository.update.mockImplementation(async (id, params) => {
    const notification = mockRepository.findOneBy(id);

    when(mockRepository.findOneBy)
      .calledWith({ id })
      .mockResolvedValue({ ...notification, ...params });
  });

  mockRepository.create.mockImplementation(params => params);
  when(mockRepository.findOneBy)
    .calledWith({ id: mockNotification.id })
    .mockResolvedValue(mockNotification);

  NotificationRepo.test(repo);
});
