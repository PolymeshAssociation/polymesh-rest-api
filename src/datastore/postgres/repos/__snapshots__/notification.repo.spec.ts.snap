// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostgresNotificationRepo Notification test suite method: create should create a notification 1`] = `
{
  "eventId": 0,
  "id": 1,
  "nonce": 1,
  "status": "active",
  "subscriptionId": 1,
  "triesLeft": 3,
}
`;

exports[`PostgresNotificationRepo Notification test suite method: findById should find the created notification 1`] = `
{
  "id": 1,
}
`;
